### PR TITLE
Fixes for ChoiceLayer sub-layers

### PR DESCRIPTION
**Pull request by patrick-wilken**
Some recent changes did not reflect sub-layers, in particular those of the ChoiceLayer in case of two targets.

Apart from these minor changes, thanks to correctly backtracking the beams of all output layers of the rec layer via _opt_search_resolve, ChoiceLayer with two targets is finally fully working in the master branch, nice! ':D

### CODE REVIEW

File: TFNetworkLayer.py [hunk 1]
```diff
@@ -2036,7 +2036,7 @@ def get_out_data_from_opts(cls, name, sources=(), extra_deps=(), out_type=None,
     :param int|None|NotSpecified n_out:
     :rtype: Data
     """
-    if out_type or n_out is not NotSpecified:
+    if out_type or (n_out is not NotSpecified and sources and n_out != sources[0].output.dim):
```
Comment (albertz) [inline TFNetworkLayer.py]: This is about `CopyLayer`. Why this change?
(Maybe actually this whole `if` branch could be removed; it's a bit historical. But anyway, it seems unrelated to the remaining PR.)

File: TFNetworkRecLayer.py [hunk 2]
```diff
@@ -2369,9 +2369,10 @@ def _opt_search_resolve(self, layer_name, acc_ta, final_net_vars):
     from TFUtil import nd_indices, assert_min_tf_version, expand_dims_unbroadcast, get_valid_scope_name_from_str
     from TFUtil import get_shape_dim
     rec_layer = self.parent_rec_layer
-    if layer_name not in self.net.layers:  # inside loop
+    layer = self.net.get_layer(layer_name)
+    if not layer:  # inside loop
```
Comment (albertz) [inline TFNetworkRecLayer.py]: This looks wrong. `self.net.get_layer` never returns `None`. If the layer does not exists, it would throw a `LayerNotFound` exception.

File: TFNetworkRecLayer.py [hunk 3]
```diff
@@ -3049,7 +3050,7 @@ def _get_cell(self):
       return self._cell
     rec_layer = self.network.parent_layer
     assert isinstance(rec_layer, RecLayer)
-    cell = self.network.parent_layer
+    cell = self.network.parent_layer.cell
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Whoops. How was it actually possible that we/I did not run into this bug before?

File: TFNetworkRecLayer.py [hunk 4]
```diff
@@ -4501,12 +4502,23 @@ def get_sub_layer_out_data_from_opts(cls, layer_name, parent_layer_kwargs):
...
+    parent_layer_name = parent_layer_kwargs["name"]
+    network = parent_layer_kwargs["network"]
+
     # The sub-layer with index n will output the n-th target. The out_data is taken directly
     # from the target as it is done in self.get_out_data_from_opts().
-    sub_layer_out_data = cls.get_out_data_from_opts(target=targets[index], network=parent_layer_kwargs["network"],
+    sub_layer_out_data = cls.get_out_data_from_opts(name=parent_layer_name + "/" + layer_name,
+                                                    sources=parent_layer_kwargs["sources"],
+                                                    target=targets[index],
+                                                    network=network,
```
Comment (albertz) [inline TFNetworkRecLayer.py]: Don't indent code like this. Make a newline and the indent by 2 spaces.

File: TFNetworkRecLayer.py [hunk 5]
```diff
@@ -4501,12 +4502,23 @@ def get_sub_layer_out_data_from_opts(cls, layer_name, parent_layer_kwargs):
...
                                                     beam_size=parent_layer_kwargs["beam_size"])
+
+    if network.search_flag:
+        # Set name of beam to parent_layer, so that they compare to equal.
+        sub_layer_out_data.beam.name = "%s%s" % (network.get_absolute_name_prefix(), parent_layer_name)
```
Comment (albertz) [inline TFNetworkRecLayer.py]: We indent by 2 spaces.

Comment (albertz) [reply TFNetworkRecLayer.py]: Btw, regarding the code: Instead of overwriting the `beam.name`, maybe better resetting the `beam` itself, to `parent_layer.beam` or so? I feel that this is a bit ugly, and has certain assumptions hard-coded about the name of the beam, which might easily break later.

File: TFNetworkLayer.py (hunk 1 above)
Comment (patrick-wilken) [reply TFNetworkLayer.py]: Sorry, I actually made some comments on Friday, but somehow they weren't posted.

I ran into the problem that `n_out` is set via `transform_config_dict` if the layer has a target. Using `super(CopyLayer, cls).get_out_data_from_opts` in this case fails to set `sparse=True`. (This happened copying from a layer in the rec unit to outside, so it's somewhat related.)

With this, I only use the base method if `n_out` is "unexpected", i.e. different than what get_concat_sources_data_template would result in.

I would be happy to remove the `if`. Because in a CopyLayer we should always know the shape from the inputs, so I don't see when setting `out_type` should be necessary.

File: TFNetworkRecLayer.py (hunk 3 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: We always set the cell when creating a TemplateLayer in `GetLayer.__call__()`, such that `self._cell` is returned. I didn't do that in `_TemplateLayer.get_sub_layer` (I guess I should/could have...).

File: TFNetworkRecLayer.py (hunk 5 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I agree this is not clean. But `get_sub_layer_out_data_from_opts()` has no access to the parent layer (as it's a classmethod), only its arguments.
I think I will create a new `SearchBeam` object in this place using the parent layer name, this should be better.

File: TFNetworkLayer.py (hunk 1 above)
Comment (albertz) [reply TFNetworkLayer.py]: Anyway. This seems unrelated to this pull request. So don't mix it in here.

But maybe create a new pull request, where you remove the whole `if` branch, and then we can see if any test case breaks.

**Comment by patrick-wilken** [comment]
Ok, I removed the CopyLayer change and tried to address the other issues.

File: TFNetworkRecLayer.py (hunk 2 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: Did you resolve this? Can you comment?

File: TFNetworkRecLayer.py [hunk 6]
```diff
@@ -4428,6 +4430,22 @@ def transform_config_dict(cls, d, network, get_layer):
         [get_layer(name) for name in d["explicit_search_sources"]] if network.search_flag else [])
     super(ChoiceLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
 
+  @classmethod
+  def _create_search_beam(cls, name, beam_size, sources, network):
+    """
+    :param str name:
+    :param int beam_size:
+    :param list[LayerBase] sources:
+    :param TFNetwork.TFNetwork network:
```
Comment (albertz) [inline TFNetworkRecLayer.py]: `:rtype:` not specified.

File: TFNetworkRecLayer.py [hunk 7]
```diff
@@ -4428,6 +4430,22 @@ def transform_config_dict(cls, d, network, get_layer):
...
+    """
+    from TFUtil import SearchBeam
+    search_dep = NotSpecified
+    if sources and sources[0] and not sources[0].output.undefined:
+      search_dep = sources[0].output.beam
+    return SearchBeam(
+      beam_size=beam_size, dependency=search_dep,
+      name="%s%s" % (network.get_absolute_name_prefix(), name))
```
Comment (albertz) [inline TFNetworkRecLayer.py]: This has still the hardcoded solution. This is not nice and can potentially break later.

File: TFNetworkRecLayer.py (hunk 2 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Yes, it was wrong. I'm now catching `LayerNotFound` exceptions and return the original tensor array in the except block.

File: TFNetworkRecLayer.py (hunk 7 above)
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Uhm, what exactly do you mean is hardcoded here? The fact that the same SearchBeam is created twice, relying on the name? 

Well. the idea of `get_sub_layer_out_data_from_opts` was to create `out_data` of the sub-layer without depending on the `out_data` or an instance of the parent layer. Do you want to change that?

Another option would be to create the parent layer beam by calling `get_out_data_from_opts` using the `parent_layer_kwargs` and then just use `out_data.beam` and discard the rest. But this is the most generic you could do here, right?

Comment (albertz) [reply TFNetworkRecLayer.py]: No, for `name` you assume that it matches to other code.

Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I'm still not sure I get what you mean. An instance of SearchBeam is characterised by the owner layer name and comparison of beams relies on the name, this is nothing I introduced. Also I did not change the name format `"%s%s" % (network.get_absolute_name_prefix(), name))`.
All I need to do is to use `parent_layer_kwargs["name"]` instead of `layer_name` when creating the beam of the sub layer. How does this depend on other code? Well, I'm assuming that the parent layer beam has indeed the name of the parent layer, but that's it.

Comment (albertz) [reply TFNetworkRecLayer.py]: You assume that we have this specific name format.

Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Yes, because it was there before ':D As of now, it is a local property of ChoiceLayer. And the class methods have access to it.
I want to go from kwargs to SearchBeam. How would I do this without using some kind of naming convention?

Comment (albertz) [reply TFNetworkRecLayer.py]: I'm just saying that you have this assumption here about the name format, and this could break at some point. Better write code in a way that it will not break.

Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: I really don't understand what my PR changed about the naming. As before, there is one, and only one, place in the code that *defines* the naming of SearchBeams.
Again, how do I get from kwargs to SearchBeam? In kwargs there is no SearchBeam, so I have to create one. For this I need a name.

Comment (albertz) [reply TFNetworkRecLayer.py]: I thought there are two places now? Where was the `SearchBeam` created before?

File: TFNetworkRecLayer.py [hunk 8]
```diff
@@ -4457,13 +4475,7 @@ def get_out_data_from_opts(cls, name, sources, target, network,
...
       out_data = Data(name="%s_output" % name, shape=shape, sparse=True, dim=out_data.dim)
     if search:
-      from TFUtil import SearchBeam
-      search_dep = NotSpecified
-      if sources and sources[0] and not sources[0].output.undefined:
-        search_dep = sources[0].output.beam
-      out_data.beam = SearchBeam(
-        beam_size=beam_size, dependency=search_dep,
-        name="%s%s" % (network.get_absolute_name_prefix(), name))
+      out_data.beam = cls._create_search_beam(name, beam_size, sources, network)
```
Comment (albertz) [inline TFNetworkRecLayer.py]: You should always use named kwargs if possible, if you have more than 1 param or there can be any ambiguity from the function name about the params.

File: TFNetworkRecLayer.py [hunk 9]
```diff
@@ -4457,13 +4475,7 @@ def get_out_data_from_opts(cls, name, sources, target, network,
       del shape[out_data.batch_dim_axis]
       out_data = Data(name="%s_output" % name, shape=shape, sparse=True, dim=out_data.dim)
     if search:
-      from TFUtil import SearchBeam
-      search_dep = NotSpecified
-      if sources and sources[0] and not sources[0].output.undefined:
-        search_dep = sources[0].output.beam
-      out_data.beam = SearchBeam(
-        beam_size=beam_size, dependency=search_dep,
-        name="%s%s" % (network.get_absolute_name_prefix(), name))
```
[the remark this answers is not in the document]
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Ah, ok, sorry ':D In general comments get a bit messy when I force push. Should I better commit fixups and then rebase when approved?

[the remark this answers is not in the document]
Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: Or is it the same with fixups...?

**Comment by patrick-wilken** [comment]
Added result type and using kwargs now.

File: TFNetworkRecLayer.py (hunk 8 above)
Comment (albertz) [reply TFNetworkRecLayer.py]: So, is this resolved? (If you resolve sth, and the old code gets outdated, I cannot see the new code, so I don't directly see it. Please click yourself on "resolve".)

File: TFNetworkRecLayer.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (albertz) [reply TFNetworkRecLayer.py]: There is no difference in the final diff (here shown on GitHub), no matter if you rebase (force push) or just add commits. So this has no effect.
It's just hard in general to see from a diff (or easy to miss), that you changed the only place where `SearchBeam` was created.

File: TFNetworkRecLayer.py [hunk 10]
```diff
@@ -4428,6 +4430,23 @@ def transform_config_dict(cls, d, network, get_layer):
...
     super(ChoiceLayer, cls).transform_config_dict(d, network=network, get_layer=get_layer)
 
+  @classmethod
+  def _create_search_beam(cls, name, beam_size, sources, network):
+    """
+    :param str name:
+    :param int beam_size:
+    :param list[LayerBase] sources:
+    :param TFNetwork.TFNetwork network:
+    :rtype: SearchBeam
```
Comment (albertz) [inline TFNetworkRecLayer.py]: I think the type is wrong, if it is not globally imported. (Did you check the type inference for the function?) I guess it must be `TFUtil.SearchBeam`.

Comment (patrick-wilken) [reply TFNetworkRecLayer.py]: True, fixed now.

**Comment by albertz** [comment]
Merged now. Maybe it makes sense to add a simple short test case for this later.
